jax/_src/scipy/special.py

```diff
@@ -657,3 +659,92 @@ def i1e(x):
 def i1(x):
   x, = _promote_args_inexact("i1", x)
   return lax.mul(lax.exp(lax.abs(x)), lax.bessel_i1e(x))
+
+def lpmn(m, n, z):
+  r"""The normalized associated Legendre functions of the first kind.
+
+  In the descriptions of the associated Legendre functions (ALFs), we use `l`
+  to denote the degree (`l_max` is the maximum degree) and `m` to denote
+  the order. The ALFs of the first kind are often used in spherical harmonics.
+  The spherical harmonic of degree `l` and order `m` can be written as
+  .. math::
+    \begin{equation}
+      Y_l^m (\theta, \phi) = N_l^m P_l^m(\cos \theta) e^{i m \phi},
+    \end{equation}
+  where
+  .. math::
+    \begin{equation}
+      N_l^m = \sqrt{\displaystyle \frac{\left(2l + 1\right)
+              \left(l-m\right)!}{4 \pi \left(l+m\right)!}}
+    \end{equation}
+  represents the normalization constant, and θ and φ are the colatitude and
+  longitude, respectively. :math:`N_l^m` is chosen in the way that the spherical
+  harmonics form a set of orthonormal basis function of :math:`L^2(S^2)`.
+  The normalized associated Legendre functions can be expressed as
+  .. math::
+    \begin{equation}
+      \bar{P}_l^m = N_l^m P_l^m.
+    \end{equation}
+
+  Args:
+    m: The maximum order of the associated Legendre functions.
+    n: The maximum degree of the associated Legendre function, often called
+      `l` in describing ALFs. Both the degrees and orders are
+      `[0, 1, 2, ..., l_max]`.
+    z: A vector of type `float32` and `float64` containing the sampling
+      points in spherical coordinates, at which the ALFs are computed;
+      and `z` is essentially `cos(θ)`.
+
+  Returns:
+    The 3D array of shape `(l_max + 1, l_max + 1, len(z))` containing the
+    normalized values of the ALFs at `z`. The dimensions follow the order
+    of order `m`, degree `l`, and `z`.
+
+  Raises:
+    TypeError if type of `z` is not in (float32, float64).
+    ValueError if `z` is not 1D.
+    NotImplementedError if `m!=n`.
+  """
+  dtype = lax.dtype(z)
+  if dtype not in (jnp.float32, jnp.float64):
+    raise TypeError(
+        'x.dtype={} is not supported, see docstring for supported types.'
+        .format(dtype))
+
+  if z.ndim != 1:
+    raise ValueError('x must be a 1D array.')
+
+  if m != n:
+    raise NotImplementedError('Computations for m!=n are not yet supported.')
+
+  l_max = n
+  p = np.zeros((l_max + 1, l_max + 1, z.shape[0]))
+
+  # The initial value p(0,0).
+  initial_value = 0.5 / np.sqrt(math.pi)
+  p[0, 0] = initial_value
+
+  # Compute the diagonal entries p(l,l) with recurrence.
+  y = np.sqrt(1.0 - z * z)
+  for l_idx in range(1, l_max + 1):
+    a = -1.0 * np.sqrt(1.0 + 0.5 / l_idx)
+    p[l_idx, l_idx] = a * y * p[l_idx - 1, l_idx - 1]
```

Comment:
Something is strange here... `p` is a numpy array, but this statement populates it with a potentially traced value (`y` derives from `z`). I don't think this will work correctly with JAX transforms.

Reply:
I have a similar concern to @jakevdp. If this function is returning a NumPy array, why would a user use it rather than the SciPy version?

Reply:
Thanks for the comments! The proposed function has two advantages over the scipy version (1) performance and (2) numerical stability. 

The following numbers come from numerical experiments
(1) performance: 

Maximum degree | Number of spatial samplings | proposed (seconds) | scipy (seconds) | speed-up
                       128 |                                         1024 |                          0.164 |                   0.619 |         3.78x
                      256 |                                          1024 |                           0.57 |                     2.53 |         4.41x
                       512 |                                         1024 |                            2.33 |                   10.91 |         4.68x
                    1024 |                                          1024 |                          10.18 |                  48.39 |          4.75x

(2) numerical stability
setting maximum degree to 512 and number of sampling points to 1024

order = 128, degree = 256, sampling point index = 1024

proposed = mode 0.00047938111886318223
scipy = 5.430959668839146e+301